File: Data Structures & Algorithms/house-robber/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        cache = [-1] * len(nums)
        def dp(i):
            if i <= -1:
                return 0
                
            if cache[i] != -1:
                return cache[i]
                
            
            cache[i] = max(dp(i-1), nums[i] + dp(i-2))
            return cache[i]
        
        return dp(len(nums)-1)

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-4.java
class Solution {
    public int maxProfit(int[] prices) {
        int buy = 0;
        int sell = 1;

        int maxP = 0;

        while(sell < prices.length){
            
            if(prices[sell] > prices[buy]){

                int profit = prices[sell] - prices[buy];
                maxP = Math.max(profit, maxP);
            }

            else{
                buy = sell;
            }

            sell++;
            
        }

        return maxP;
    }
}
